### PR TITLE
Better documentation

### PR DESCRIPTION
**Pull request by Kukunin**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,7 +14,15 @@
 [![Test Coverage](https://codeclimate.com/github/solnic/transproc/badges/coverage.svg)][codeclimate]
 [![Inline docs](http://inch-ci.org/github/solnic/transproc.svg?branch=master)][inchpages]
 
-Transproc is a small library that allows you to compose methods into a functional pipeline using left-to-right function composition. It works like `|>` in Elixir or `>>` in F#.
+Transproc is a small library that allows you to compose methods into a functional
```
Comment (AMHOL) [inline README.md]: Not necessarily `methods` that are composed, I'd probably just go with `compose procs into a functional transformation`

File: README.md [hunk 2]
```diff
@@ -14,7 +14,15 @@
...
+pipeline using left-to-right function composition.
+
+The approach came from Functional Programming, where small functions are composed
```
Comment (AMHOL) [inline README.md]: Not sure about `small functions are composed into bigger functions`, how about `simple functions are composed into more complex functions` or something along those lines?

File: README.md [hunk 3]
```diff
@@ -14,7 +14,15 @@
...
+into bigger functions in order to transform some data. It works like `|>` in Elixir
+or `>>` in F#.
+
+`transproc` provides a mechanism to define and composite transformations, along
```
Comment (AMHOL) [inline README.md]: Should this be `transproc provides a mechanism to define composite transformations` or `transproc provides a mechanism to define and compose transformations`?

File: README.md [hunk 4]
```diff
@@ -34,7 +42,110 @@ Or install it yourself as:
 
     $ gem install transproc
 
-## Usage
+## Basics
+
+The simple transformations are defined as easy as
```
Comment (AMHOL) [inline README.md]: Would drop `The` from the start of this sentence and just go with `Simple transformations are defined as easy as:`

File: README.md [hunk 5]
```diff
@@ -34,7 +42,110 @@ Or install it yourself as:
...
+```
+
+It's easy to compose transformations
+
+```ruby
+to_string = Transproc::Function.new(:to_s.to_proc)
+(increament >> to_string)[1] => '2'
+```
+
+It's easy to use extra arguments to transformation
```
Comment (AMHOL) [inline README.md]: Would go with `It's easy to pass extra arguments to transformations:`

File: README.md [hunk 6]
```diff
@@ -34,7 +42,110 @@ Or install it yourself as:
...
+
+```ruby
+append = Transproc::Function.new(-> (value, suffix) { value + suffix })
+append_bar = append.with('_bar')
+append_bar['foo'] # => foo_bar
+```
+
+Or even accept other transformation as argument
```
Comment (AMHOL) [inline README.md]: Should either be `Or even accept transformations as arguments:` or `Or even accept another transformation as an argument:`

File: README.md [hunk 7]
```diff
@@ -34,7 +42,110 @@ Or install it yourself as:
...
+  def self.map_array(array, fn)
+    array.map(&fn)
+  end
+end
+M[:map_array, M[:to_string]].([1, 2, 3]) # => ['1', '2', '3']
+```
+
+### Built-in transformations
+
+`transproc` comes with a lot of built-in functions. They come in form of module
```
Comment (AMHOL) [inline README.md]: Should be `They come in the form of modules with class methods, which you can import into a registry.

File: README.md [hunk 8]
```diff
@@ -34,7 +42,110 @@ Or install it yourself as:
...
+
+* [Coercions](http://www.rubydoc.info/gems/transproc/Transproc/Coercions)
+* [Array transformations](http://www.rubydoc.info/gems/transproc/Transproc/ArrayTransformations)
+* [Hash transformations](http://www.rubydoc.info/gems/transproc/Transproc/HashTransformations)
+* [Class transformations](http://www.rubydoc.info/gems/transproc/Transproc/ClassTransformations)
+* [Proc transformations](http://www.rubydoc.info/gems/transproc/Transproc/ProcTransformations)
+* [Conditional](http://www.rubydoc.info/gems/transproc/Transproc/Conditional)
+* [Recursion](http://www.rubydoc.info/gems/transproc/Transproc/Recursion)
+
+Load everything with
```
Comment (AMHOL) [inline README.md]: Think this could be updated to:
----------------------
You can import everything with:

```ruby
module T
  extend Transproc::Registry

  import Transproc::Coercions
  import Transproc::ArrayTransformations
  import Transproc::HashTransformations
  import Transproc::ClassTransformations
  import Transproc::ProcTransformations
  import Transproc::Conditional
  import Transproc::Recursion
end
T[:to_string].call(:abc) # => 'abc'
```

Or import selectively with:

```ruby
module T
  extend Transproc::Registry

  import :to_string, from: Transproc::Coercions, as: :stringify
end
T[:stringify].call(:abc) # => 'abc'
T[:to_string].call(:abc)
# => Transproc::FunctionNotFoundError: No registered function T[:to_string]
```

File: README.md [hunk 9]
```diff
@@ -34,7 +42,110 @@ Or install it yourself as:
...
+  import Transproc::ProcTransformations
+  import Transproc::Conditional
+  import Transproc::Recursion
+end
+T[:to_string].call(:abc) # => 'abc'
+```
+
+### Transformers
+
+Transformers is class level DSL for mappers. For example:
```
Comment (AMHOL) [inline README.md]: Perhaps go with singular here and for the heading, something like:

`Transformer is a class-level DSL for composing transformation pipelines, for example:`

File: README.md [hunk 10]
```diff
@@ -34,7 +42,110 @@ Or install it yourself as:
...
+    map_value :address do
+      constructor_inject Test::Address
+    end
+    constructor_inject Test::User
+  end
+end
+```
+
+It converts every method call to correspondent transformation, and joins sib
+method calls into sequence of transformations - transproc.
```
Comment (AMHOL) [inline README.md]: `It converts every method call to its corresponding transformation, and joins these transformations into a transformation pipeline (a transproc).`

File: README.md [hunk 11]
```diff
@@ -34,7 +42,110 @@ Or install it yourself as:
...
+
+For every given block transformer evaluates it in nested scope, and passes
+converted transproc as last argument to transformation.
```
Comment (AMHOL) [inline README.md]: I think we can probably remove this part as it's just an implementation detail.

Comment (Kukunin) [inline README.md]: This is only one item I'd comment.
I think we still need to leave this paragraph, maybe rephrased. 

For me as for `transproc` beginner this information was crucial to understand in order to know how to use transformer DSL. 

Also, one should know a signature of every transformation before use, so this fact makes sense to know as well.

**Comment by solnic** [comment]
This LGTM, @AMHOL ?

**Comment by solnic** [comment]
uhm sorry I didn't notice that "AMHOL approved these changes a day ago" (gotta get used to this new review feature :))

Just merged this, thank you @Kukunin this is very helpful!
